charms/jupyter-controller/reactive/jupyter_controller.py

```diff
@@ -18,7 +18,11 @@ def charm_ready():
     layer.status.active('')
 
 
-@when_any('layer.docker-resource.oci-image.changed', 'config.changed')
+@when_any(
+    'layer.docker-resource.oci-image.changed',
+    'config.changed',
+    'endpoint.service-mesh.joined',
```

Comment:
This will trigger on any hook as long as the relation is joined.  I think I'd recommend `endpoint.service-mesh.changed` instead and [make sure you clear it manually](https://charmsreactive.readthedocs.io/en/latest/managed-flags.html#id16).